### PR TITLE
array: generalize S.take and S.drop

### PR DESCRIPTION
**Pull request by davidchambers**
The code is more complex than I would like it to be, but I believe the complexity is necessary. Please correct me if I'm wrong about this! :)

### CODE REVIEW

**Comment by davidchambers** [comment]
> I think it should be possible to do it in one pass

Could you elaborate, Brad? I believe we're traversing the structure just once, with `Z.reduce`.

File: index.js [hunk 1]
```diff
@@ -2928,31 +2928,55 @@
...
-  //. > S.take (5) (['a', 'b', 'c', 'd', 'e'])
-  //. Just (['a', 'b', 'c', 'd', 'e'])
+  //. > S.take (1) (['foo', 'bar'])
+  //. Just (['foo'])
   //.
-  //. > S.take (6) (['a', 'b', 'c', 'd', 'e'])
+  //. > S.take (2) (['foo', 'bar'])
+  //. Just (['foo', 'bar'])
+  //.
+  //. > S.take (3) (['foo', 'bar'])
```
[the remark this answers is not in the document]
Comment (davidchambers) [reply index.js]: Good idea. I've added an example to each section. :)

**Comment by Bradcomp** [comment]
Ugh, I am tired today, but I see what is happening now. `reduce` is returning a `Just Pair (Int (f a))`, and we are rejecting if the int is > 0 because it means the foldable wasn't long enough. Otherwise we take the resultant array.

I had things inside out in my thinking.

Looks good to me!

**Comment by masaeedu** [comment]
Hi @davidchambers. I talked to @Bradcomp for a bit about this PR, I think maybe we don't really need the `Applicative` constraint on here.

**Comment by davidchambers** [comment]
The Applicative constraint is not obvious. It does not arise from `_takeDrop` itself, but from the `generalCase` functions provided to `_takeDrop` in the definitions of `take` and `drop`:

```javascript
var take = _takeDrop (
  function(n, xs) { return xs.slice (0, n); },
  function(n, xs, x) { return n > 0 ? Z.append (x, xs) : xs; }
  //                                  ^^^^^^^^^^^^^^^^
);
```

```javascript
var drop = _takeDrop (
  function(n, xs) { return xs.slice (n); },
  function(n, xs, x) { return n > 0 ? xs : Z.append (x, xs); }
  //                                       ^^^^^^^^^^^^^^^^
);
```

**Pull request merged**